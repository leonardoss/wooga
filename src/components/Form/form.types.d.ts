export interface ValidationType {
  [key: string]: {
    value: string;
    msg: string;
  };
}
export interface FieldType {
  type: string;
  label: string;
  name: string;
  validation?: ValidationType;
}
export interface FormProps {
  action: string;
  fields: FieldsType[];
  layout?: string;
  method?: string;
  submitButton?: SubmitButtonType;
  onSubmitForm: (any) => void;
}
export interface SubmitButtonType {
  label?: string;
}
export type GenericObjectType = { [key: string]: any };
export type State = Record<string, unknown>;
export type SetState = (prevState: GenericObjectType) => { [x: string]: string }