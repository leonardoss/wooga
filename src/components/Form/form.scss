//generics
*,
*:before,
*:after {
  box-sizing: border-box;
}
.btn {
  background: #27343c;
  border: none;
  display: block;
  height: 60px;
  padding: 0 40px;
  color: #fff;
  font-size: 20px;
  margin: auto;
}
.input-text {
  height: 60px;
  width: 100%;
  font-size: 20px;
  padding: 0 0 0 10px;
  border: 1px solid #ccc;
}

.form-control {
  margin-bottom: 12px;
}
.form-label {
  font-size: 20px;
}
.error {
  color: red;
}
.layout-horizontal {
  display: flex;
  align-items: flex-end;
}
