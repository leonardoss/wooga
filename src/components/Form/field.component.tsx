import React, { FC } from 'react';

import { FieldType, GenericObjectType } from './form.types';

interface FieldProps {
  field: FieldType;
  setFieldValue: (arg: GenericObjectType) => void;
  fieldValue: GenericObjectType;
  errors: GenericObjectType;
}

const Field: FC<FieldProps> = ({ field, setFieldValue, fieldValue, errors }: FieldProps) => {
  const { name, label, type } = field;

  return (
    <div className="form-control" key={name}>
      <label className="form-label" htmlFor={name}>
        {label || name}
      </label>
      <input
        type={type || 'text'}
        name={name}
        className={`input-${type}`}
        value={fieldValue[name] || ''}
        onChange={({ target }) =>
          setFieldValue((prevState: GenericObjectType) => ({
            ...prevState,
            [target.name]: target.value,
          }))
        }
      />
      {errors[name] &&
        Object.keys(errors[name]).map((key) => (
          <div key={errors[name][key]} className="error">
            * {errors[name][key]}
          </div>
        ))}
    </div>
  );
};

export default Field;
