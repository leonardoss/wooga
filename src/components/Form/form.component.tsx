import React, { FC, FormEvent, useState } from 'react';

import Field from './field.component';

import { FormProps, FieldType, GenericObjectType, ValidationType } from './form.types';
import type { State } from './form.types';

import './form.scss';

export const validateField = (key: string, value: string, currentValue: string) => {
  switch (key) {
    case 'required':
      return currentValue !== '';
    case 'fieldStartsWith':
      return currentValue.includes(value);
  }
};

const Form: FC<FormProps> = ({
  fields,
  action,
  method,
  layout,
  onSubmitForm,
  submitButton,
}: FormProps) => {
  const [fieldValue, setFieldValue] = useState<GenericObjectType>({});
  const [errors, setErrors] = useState<any>({});

  const handleValidation = (): boolean => {
    let isValid = true;

    fields.forEach(({ name, validation }) => {
      const currentValue: string = fieldValue[name] || '';

      if (typeof validation !== 'undefined') {
        validation.map((item: ValidationType) => {
          for (const key in item) {
            if (!validateField(key, item[key].value, currentValue)) {
              setErrors((prevState: State) => {
                const newValue: Partial<State> = {
                  [name]: {
                    ...(prevState[name] as State),
                    [key]: `${item[key].msg}`,
                  },
                };
                return Object.assign({}, prevState, newValue);
              });

              isValid = false;
            } else {
              setErrors((prevState: State) => {
                const objTransformed: any = prevState[name];
                delete objTransformed[key];
                const newValue: Partial<State> = {
                  [name]: {
                    ...objTransformed,
                  },
                };
                return Object.assign({}, prevState, newValue);
              });
            }
          }
        });
      }
    });

    return isValid;
  };

  const handleSubmit = (event: FormEvent<EventTarget>) => {
    event.preventDefault();
    if (handleValidation()) {
      onSubmitForm(fieldValue);
    }
  };

  return (
    <form action={action} method={method} onSubmit={handleSubmit} className={`layout-${layout}`}>
      {fields.map((field: FieldType) => (
        <Field
          key={field.name}
          field={field}
          errors={errors}
          setFieldValue={setFieldValue}
          fieldValue={fieldValue}
        />
      ))}
      <div className="form-control">
        <input type="submit" value={submitButton?.label || 'Submit'} className="btn btn-submit" />
      </div>
    </form>
  );
};

export default Form;
