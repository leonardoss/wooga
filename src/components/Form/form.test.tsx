import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import Form, { validateField } from './form.component';

const options = {
  action: '/foobar',
  method: 'post',
  layout: 'vertical',
  fields: [
    {
      type: 'text',
      label: 'Name',
      name: 'someName',
      validation: [
        {
          required: {
            value: true,
            msg: 'field required',
          },
        },
        {
          fieldStartsWith: {
            value: 'wooga.name',
            msg: 'expected field starting with: wooga.name',
          },
        },
      ],
    },
    {
      type: 'text',
      label: 'Name 2',
      name: 'someName 2',
      validation: [
        {
          required: {
            value: true,
            msg: 'field required',
          },
        },
      ],
    },
  ],
  submitButton: {
    label: 'Send data',
  },
  onSubmitForm: (values: any) => {
    console.log('--+ FORM SUBMITED: values on form: ', values);
  },
};
const MOCK_PROPS = {
  fields: options.fields,
  action: '/foo/bar.php',
  method: 'GET',
  onSubmitForm: jest.fn(),
};

describe('Testing App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Form {...MOCK_PROPS} />, div);
  });
  it('Form renders correctly', () => {
    const tree = renderer.create(<Form {...MOCK_PROPS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('should chekck if "required" condition works as expected depending of the currentValue', () => {
    expect(validateField('required', '', '')).toBe(false);
    expect(validateField('required', '', 'some value')).toBe(true);
  });
  it('should chekck if "fieldStartsWith" condition works as expected depending of the currentValue', () => {
    expect(validateField('fieldStartsWith', 'some value', 'other value')).toBe(false);
    expect(validateField('fieldStartsWith', 'some value', 'some value')).toBe(true);
    expect(validateField('fieldStartsWith', 'some value', 'some value extended')).toBe(true);
  });

});
