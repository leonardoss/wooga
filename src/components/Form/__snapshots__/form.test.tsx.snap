// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing App Form renders correctly 1`] = `
<form
  action="/foo/bar.php"
  className="layout-undefined"
  method="GET"
  onSubmit={[Function]}
>
  <div
    className="form-control"
  >
    <label
      className="form-label"
      htmlFor="someName"
    >
      Name
    </label>
    <input
      className="input-text"
      name="someName"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-control"
  >
    <label
      className="form-label"
      htmlFor="someName 2"
    >
      Name 2
    </label>
    <input
      className="input-text"
      name="someName 2"
      onChange={[Function]}
      type="text"
      value=""
    />
  </div>
  <div
    className="form-control"
  >
    <input
      className="btn btn-submit"
      type="submit"
      value="Submit"
    />
  </div>
</form>
`;
